sys/analog_util/adc_util.c

```diff
@@ -14,28 +15,52 @@
  * @brief       ADC utility function implementation
  *
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Joakim Nohlgård <joakim.nohlgard@eistec.se>
  *
  * @}
  */
 
+#include <stdint.h>
+#include <inttypes.h>
+#include "cpu.h"
+#include "periph/adc.h"
 #include "analog_util.h"
 
-/* keep a max value to ADC resolution mapping for quick access in the ROM */
-static const int val_max[] = {
-    [ADC_RES_6BIT] =    0x003f,
-    [ADC_RES_8BIT] =    0x00ff,
-    [ADC_RES_10BIT] =   0x03ff,
-    [ADC_RES_12BIT] =   0x0fff,
-    [ADC_RES_14BIT] =   0x3fff,
-    [ADC_RES_16BIT] =   0xffff
-};
+#define ENABLE_DEBUG 0
+#include "debug.h"
 
-int adc_util_map(int sample, adc_res_t res, int min, int max)
+/**
+ * @brief Convert adc_res_t resolution setting into numeric bit count
+ */
+static unsigned int _adc_res_bits(adc_res_t res)
+{
+    switch (res) {
+        case ADC_RES_6BIT:
+            return 6;
+        case ADC_RES_8BIT:
+            return 8;
+        case ADC_RES_10BIT:
+            return 10;
+        case ADC_RES_12BIT:
+            return 12;
+        case ADC_RES_14BIT:
+            return 14;
+        case ADC_RES_16BIT:
+        default:
+            return 16;
```

Comment:
does this make sense, wouldn't `-1` as default be more appropriate - just in case some weird CPU defines an `adc_res_t` with values not expected here?

Reply:
We are using the result to perform a shift, and shifting by negative number of bits is undefined behaviour in C99. It will only work if we also add error returns for the adc_map functions. I know that the cc2538 periph adc driver provides some weird resolutions, like 7 bit etc. But right now I just want to focus on getting this working again so it can be used for the general use cases. I could add an always failing assert on the default instead, if that is better?